watcher/portfs.cpp

```diff
@@ -0,0 +1,323 @@
+  auto f = watchman::make_unique<watchman_port_file>();
+
+  f->name = name;
+  f->port_file.fo_name = (char*)name.c_str();
+  f->port_file.fo_atime = finfo.atime;
+  f->port_file.fo_mtime = finfo.mtime;
+  f->port_file.fo_ctime = finfo.ctime;
+  f->is_dir = finfo.isDir();
+
+  return f;
+}
+
+PortFSWatcher::PortFSWatcher(w_root_t* root)
+    : Watcher("portfs", 0),
+      port_fd(port_create(), "port_create()"),
+      port_delete_fd(port_create(), "port_create()"),
+      root_deleted(false) {
+  auto wlock = port_files.wlock();
+  wlock->reserve(root->config.getInt(CFG_HINT_NUM_DIRS, HINT_NUM_DIRS));
+  port_fd.setCloExec();
+  port_delete_fd.setCloExec();
+}
+
+bool PortFSWatcher::do_watch(
+    const w_string& name,
+    const watchman::FileInformation& finfo,
+    bool throw_on_error) {
+  auto wlock = port_files.wlock();
+  if (wlock->find(name) != wlock->end()) {
+    // Already watching it
+    return true;
+  }
+
+  auto f = make_port_file(name, finfo);
+  auto rawFile = f.get();
+  wlock->emplace(name, std::move(f));
+
+  w_log(W_LOG_DBG, "watching %s\n", name.c_str());
+  errno = 0;
+  if (port_associate(
+          port_fd.fd(),
+          PORT_SOURCE_FILE,
+          (uintptr_t)&rawFile->port_file,
+          WATCHMAN_PORT_EVENTS,
+          (void*)rawFile)) {
+    int err = errno;
+    w_log(
+        W_LOG_ERR,
+        "port_associate %s %s\n",
+        rawFile->port_file.fo_name,
+        strerror(errno));
+    wlock->erase(name);
+    if (throw_on_error) {
+      throw std::system_error(err, std::generic_category(), "port_associate");
+    }
+    return false;
+  }
+
+  return true;
+}
+
+/*
+ * We need to have an extra port the catches the delete event on the root.
+ * The reason for this is that the creation or delete of one of the
+ * files/directories directly under the root will cause an FILE_MODIFIED,
+ * FILE_ATTRIB event on the root, because just one event can be generated
+ * per file_obj the delete event coming afterwards is not seen.
+ */
+bool PortFSWatcher::start(const std::shared_ptr<w_root_t>& root) {
+  struct stat st;
+  if (stat(root->root_path.c_str(), &st)) {
+    watchman::log(watchman::ERR, "stat failed in PortFS root delete watch");
+    root->cancel();
+    return false;
+  }
+
+  auto f = make_port_file(root->root_path, watchman::FileInformation(st));
+  auto rawFile = f.get();
+  root_delete_w_port_file = std::move(f);
+
+  w_log(W_LOG_DBG, "watching %s\n for delete events", root->root_path.c_str());
+  errno = 0;
+  if (port_associate(
+          port_delete_fd.fd(),
+          PORT_SOURCE_FILE,
+          (uintptr_t)&rawFile->port_file,
+          0, // we only want the delete events
+          (void*)rawFile)) {
+    watchman::log(
+        watchman::ERR, "port_associate failed in PortFS root delete watch");
+    return false;
+  }
+  return true;
+}
+
+bool PortFSWatcher::startWatchFile(struct watchman_file* file) {
+  auto name = w_dir_path_cat_str(file->parent, file->getName());
```

Comment:
I broken this out from under you in https://github.com/facebook/watchman/commit/085866203c94f1d1b38eefa6f047b2eb16dbcb0d

I think you want this instead:

```
auto name = file->parent->getFullPathToChild(file->getName());
```